test/e2e/capi_test.go

```diff
@@ -259,3 +261,76 @@ var _ = Describe("Running the Cluster API E2E tests", func() {
 		})
 	}
 })
+
+var _ = Describe("When upgrading a workload cluster [K8s-Upgrade]", func() {
+	var (
+		ctx               = context.TODO()
+		identityNamespace *corev1.Namespace
+	)
+	BeforeEach(func() {
+		Expect(e2eConfig.Variables).To(HaveKey(capi_e2e.CNIPath))
+		rgName := fmt.Sprintf("capz-e2e-%s", util.RandomString(6))
+		Expect(os.Setenv(AzureResourceGroup, rgName)).NotTo(HaveOccurred())
+		Expect(os.Setenv(AzureVNetName, fmt.Sprintf("%s-vnet", rgName))).NotTo(HaveOccurred())
+
+		// Opt into Windows for versions greater than or equal to 1.22
+		Expect(e2eConfig.Variables).To(HaveKey(capi_e2e.KubernetesVersionUpgradeFrom))
+		v, err := semver.ParseTolerant(e2eConfig.GetVariable(capi_e2e.KubernetesVersionUpgradeFrom))
+		Expect(err).NotTo(HaveOccurred())
+		if v.GTE(semver.MustParse("1.22.0")) {
+			Expect(os.Setenv("WINDOWS_WORKER_MACHINE_COUNT", "2")).To(Succeed())
+			Expect(os.Setenv("K8S_FEATURE_GATES", "WindowsHostProcessContainers=true")).To(Succeed())
+		}
+
+		clientset := bootstrapClusterProxy.GetClientSet()
+		Expect(clientset).NotTo(BeNil())
+		ns := fmt.Sprintf("capz-e2e-identity-%s", util.RandomString(6))
+
+		identityNamespace, err = e2e_namespace.Create(ctx, clientset, ns, map[string]string{})
+		Expect(err).ToNot(HaveOccurred())
+
+		spClientSecret := os.Getenv(AzureClientSecret)
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      IdentitySecretName,
+				Namespace: identityNamespace.Name,
+				Labels: map[string]string{
+					clusterctl.ClusterctlMoveHierarchyLabelName: "true",
+				},
+			},
+			Type: corev1.SecretTypeOpaque,
+			Data: map[string][]byte{"clientSecret": []byte(spClientSecret)},
+		}
+		err = bootstrapClusterProxy.GetClient().Create(ctx, secret)
+		Expect(err).ToNot(HaveOccurred())
+
+		identityName := e2eConfig.GetVariable(ClusterIdentityName)
+		Expect(os.Setenv(ClusterIdentityName, identityName)).NotTo(HaveOccurred())
+		Expect(os.Setenv(ClusterIdentitySecretName, IdentitySecretName)).NotTo(HaveOccurred())
+		Expect(os.Setenv(ClusterIdentitySecretNamespace, identityNamespace.Name)).NotTo(HaveOccurred())
+
+	})
+
+	AfterEach(func() {
+		redactLogs()
+
+		Expect(os.Unsetenv(AzureResourceGroup)).NotTo(HaveOccurred())
+		Expect(os.Unsetenv(AzureVNetName)).NotTo(HaveOccurred())
+		Expect(os.Unsetenv(ClusterIdentityName)).NotTo(HaveOccurred())
+		Expect(os.Unsetenv(ClusterIdentitySecretName)).NotTo(HaveOccurred())
+		Expect(os.Unsetenv(ClusterIdentitySecretNamespace)).NotTo(HaveOccurred())
```

Comment:
need to unset windows vars

// Unset windows specific variables
				Expect(os.Unsetenv("WINDOWS_WORKER_MACHINE_COUNT")).To(Succeed())
				Expect(os.Unsetenv("K8S_FEATURE_GATES")).To(Succeed())